synapse/handlers/oidc.py

```diff
@@ -1035,6 +1139,161 @@ def _remote_id_from_userinfo(self, userinfo: UserInfo) -> str:
+        if not self._config.backchannel_logout_enabled:
+            logger.warning(
+                f"Received an OIDC Back-Channel Logout request from issuer {self.issuer!r} but it is disabled in config"
+            )
+            raise SynapseError(
+                400, "OpenID Connect Back-Channel Logout is disabled for this provider"
+            )
+
+        metadata = await self.load_metadata()
+
+        # As per OIDC Back-Channel Logout 1.0 sec. 2.4:
+        #   A Logout Token MUST be signed and MAY also be encrypted. The same
+        #   keys are used to sign and encrypt Logout Tokens as are used for ID
+        #   Tokens. If the Logout Token is encrypted, it SHOULD replicate the
+        #   iss (issuer) claim in the JWT Header Parameters, as specified in
+        #   Section 5.3 of [JWT].
+        alg_values = metadata.get("id_token_signing_alg_values_supported", ["RS256"])
+        jwt = JsonWebToken(alg_values)
+
+        # As per sec. 2.6:
+        #    3. Validate the iss, aud, and iat Claims in the same way they are
+        #       validated in ID Tokens.
+        # Which means the audience should contain Synapse's client_id and the
+        # issuer should be the IdP issuer
+        claim_options = {
+            "iss": {"values": [self.issuer]},
+            "aud": {"values": [self.client_id]},
+        }
+
+        logger.debug("Attempting to decode JWT logout_token %r", logout_token)
+
+        # Try to decode the keys in cache first, then retry by forcing the keys
+        # to be reloaded
+        jwk_set = await self.load_jwks()
+        try:
+            claims = jwt.decode(
+                logout_token,
+                key=jwk_set,
+                claims_cls=LogoutToken,
+                claims_options=claim_options,
+            )
+        except ValueError:
+            logger.info("Reloading JWKS after decode error")
+            jwk_set = await self.load_jwks(force=True)  # try reloading the jwks
+            claims = jwt.decode(
+                logout_token,
+                key=jwk_set,
+                claims_cls=LogoutToken,
+                claims_options=claim_options,
+            )
+
+        logger.debug("Decoded logout_token JWT %r; validating", claims)
+        claims.validate(leeway=120)  # allows 2 min of clock skew
+
+        # As per sec. 2.6:
+        #    4. Verify that the Logout Token contains a sub Claim, a sid Claim,
+        #       or both.
+        #    5. Verify that the Logout Token contains an events Claim whose
+        #       value is JSON object containing the member name
+        #       http://schemas.openid.net/event/backchannel-logout.
+        #    6. Verify that the Logout Token does not contain a nonce Claim.
+        # This is all verified by the LogoutToken claims class, so at this
+        # point the `sid` claim exists and is a string.
+        sid: str = claims.get("sid")
+
+        # Find in the store if we have a device associated with that SID
```

Comment:
Maybe this is clearer?

```suggestion
        # Fetch any device(s) in the store associated with the session ID.
```